### PR TITLE
[GemDBHelper] Retry when db connection error

### LINKED ISSUE #115: Retry PG timeouts rather than crashing update task
If a postgres query times out, the entire update task crashes:

```
Oct 24 08:01:01 bundler-api app/scheduler.4810:  ** Invoke update (first_time) 
Oct 24 08:01:01 bundler-api app/scheduler.4810:  ** Execute update 
Oct 24 08:01:02 bundler-api app/scheduler.4810:  # of specs from indexes: 629167 
Oct 24 08:01:08 bundler-api app/scheduler.4810:  # of non yanked local gem versions: 629166 
Oct 24 08:01:16 bundler-api app/scheduler.4810:  Finished Enqueuing Jobs! 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  Sequel::DatabaseConnectionError: PG::ConnectionBad: timeout expired 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/adapters/postgres.rb:236:in `initialize' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/adapters/postgres.rb:236:in `new' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/adapters/postgres.rb:236:in `connect' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool.rb:101:in `make_new' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:160:in `make_new' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:137:in `available' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:127:in `block in acquire' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:201:in `block in sync' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:201:in `synchronize' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:201:in `sync' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:126:in `acquire' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/connection_pool/threaded.rb:88:in `hold' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/database/connecting.rb:250:in `synchronize' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/adapters/postgres.rb:866:in `literal_string_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:78:in `literal_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:671:in `block in placeholder_literal_string_sql_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:668:in `loop' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:668:in `placeholder_literal_string_sql_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/sql.rb:107:in `to_s_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:1192:in `literal_expression_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:85:in `literal_append' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:343:in `literal' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:1508:in `static_sql' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/sql.rb:222:in `select_sql' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/actions.rb:139:in `each' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel_pg-1.6.11/lib/sequel_pg/sequel_pg.rb:58:in `each' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/actions.rb:640:in `single_record' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/vendor/bundle/ruby/2.2.0/gems/sequel-4.18.0/lib/sequel/dataset/actions.rb:192:in `first' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/lib/bundler_api/update/gem_db_helper.rb:30:in `exists?' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/lib/bundler_api/update/job.rb:17:in `run' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/lib/bundler_api/update/consumer_pool.rb:36:in `block (2 levels) in create_thread' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/lib/bundler_api/update/consumer_pool.rb:32:in `loop' 
Oct 24 08:01:37 bundler-api app/scheduler.4810:  /app/lib/bundler_api/update/consumer_pool.rb:32:in `block in create_thread' 
```

It would be better to catch the timeout exception and retry at least 2-3x before crashing.

### PR DESCRIPTION
**Pull request by allenzhao**
Fixes #115.

### CODE REVIEW

File: lib/bundler_api/update/gem_db_helper.rb [hunk 1]
```diff
@@ -25,7 +25,16 @@ def exists?(payload)
       AND versions.indexed = true
     SQL
 
-    result = dataset.first
+    begin
+      result = dataset.first
+      tries  = 2 #Retry two times if there is db connection error
+    rescue Sequel::DatabaseConnectionError => e
+      sleep
+      retry unless (tries -= 1).zero?
```
Comment (segiddins) [inline lib/bundler_api/update/gem_db_helper.rb]: `tries` will be undefined, no?

Comment (allenzhao) [inline lib/bundler_api/update/gem_db_helper.rb]: Ah, and the comment breaks CI check... I will fix that;) Thanks!

**Comment by segiddins** [comment]
This will also require a test covering the new code before we can merge it, but thanks so much for tackling this!

**Comment by allenzhao** [comment]
@segiddins Please review ;)
And btw, I found a warning in Travis saying that it doesn't provide sudo and:
`If you require sudo, add 'sudo: required' to your .travis.yml`
See [here](https://travis-ci.org/bundler/bundler-api/builds/112898602#L143-L145)
And another warning is that
`ERROR -- : Not loading: No config file found at '/home/travis/build/bundler/bundler-api/lib/bundler_api/config/appsignal.yml' and no APPSIGNAL_PUSH_API_KEY env var present`
See [here](https://travis-ci.org/bundler/bundler-api/builds/112898602#L290)

**Comment by segiddins** [comment]
Looks good, but can you verify that the sleep doesn't actually run in the tests? You may want to stub that as well

**Comment by allenzhao** [comment]
@segiddins Please review again ;)
And I rebased to make the commit history clean

**Comment by allenzhao** [comment]
@segiddins Just wondering if this can be merged or is there anything else that I need to do?
Thanks!

**Comment by segiddins** [comment]
@homu r+

**Pull request merged**